core/src/miner/backup.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright 2019 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+
+use ckey::Public;
+use kvdb::{DBTransaction, KeyValueDB};
+use primitives::H256;
+use rlp::Encodable;
+
+use super::mem_pool_types::{MemPoolItem, PoolingInstant};
+use crate::db as dblib;
+
+const PREFIX_SIZE: usize = 5;
+const ITEM: &[u8; PREFIX_SIZE] = b"item_";
+const SEQS: &[u8; PREFIX_SIZE] = b"seqs_";
+
+pub struct RecoveredData {
+    pub by_hash: HashMap<H256, MemPoolItem>,
+    pub first_seqs: HashMap<Public, u64>,
+    pub last_time: PoolingInstant,
+    pub last_timestamp: u64,
+    pub next_transaction_id: u64,
+}
+
+pub fn backup_item(batch: &mut DBTransaction, key: H256, item: &MemPoolItem) {
+    let mut db_key = ITEM.to_vec();
+    db_key.extend_from_slice(key.as_ref());
+    batch.put(dblib::COL_MEMPOOL, db_key.as_ref(), item.rlp_bytes().as_ref());
+}
+
+pub fn backup_seqs(batch: &mut DBTransaction, key: Public, value: u64) {
+    let mut db_key = SEQS.to_vec();
+    db_key.extend_from_slice(key.as_ref());
+    batch.put(dblib::COL_MEMPOOL, db_key.as_ref(), value.rlp_bytes().as_ref());
+}
+
+pub fn backup_extra(batch: &mut DBTransaction, key: &[u8], val: u64) {
+    batch.put(dblib::COL_MEMPOOL, key, val.rlp_bytes().as_ref());
+}
+
+pub fn remove_item(batch: &mut DBTransaction, key: &H256) {
+    let mut db_key = ITEM.to_vec();
+    db_key.extend_from_slice(key.as_ref());
+    batch.delete(dblib::COL_MEMPOOL, db_key.as_ref());
+}
+
+pub fn recover_to_data(db: &KeyValueDB) -> RecoveredData {
+    let mut by_hash = HashMap::new();
+    let mut first_seqs = HashMap::new();
+
+    for (key, value) in db.iter(dblib::COL_MEMPOOL) {
+        let key_prefix: &[u8] = &key.as_ref()[0..PREFIX_SIZE];
+        let bytes = (*value).to_vec();
+        let rlp = rlp::Rlp::new(&bytes);
+
+        if b"item_" == key_prefix {
+            let decoded_key = (key.as_ref()[5..]).into();
+            let decoded_item= rlp.as_val();
+            by_hash.insert(decoded_key, decoded_item);
+        } else if b"seqs_" == key_prefix {
```

Comment:
Is `if SEQS == key_prefix` not available?

Reply:
Fixed